File: examples/kotlin/examples/ClientFunctions.kt
/**
 * Functions Example for ekoDB Kotlin Client
 *
 * Demonstrates creating, managing, and executing functions with the Kotlin client.
 * Covers: FindAll, Group, Project, Count, and Function management operations.
 */

package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.ekodb.client.getValue
import io.ekodb.client.functions.UserFunction
import io.ekodb.client.functions.ParameterDefinition
import io.ekodb.client.functions.FunctionStageConfig
import io.ekodb.client.functions.GroupFunctionConfig
import io.ekodb.client.functions.GroupFunctionOp
import io.ekodb.client.functions.QueryConditionOperator
import io.ekodb.client.functions.queryCondition
import io.ekodb.client.types.Record
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

private const val TEST_COLLECTION = "client_functions_users_kt"
private const val ACTIVE_LABEL = "client_functions_get_active_users_kt"
private const val STATUS_LABEL = "client_functions_get_users_by_status_kt"
private const val STATS_LABEL = "client_functions_user_stats_kt"
private const val TOP_LABEL = "client_functions_top_users_kt"
private const val COUNT_LABEL = "client_functions_count_users_kt"

/**
 * Detect the server's "label already exists" rejection.
 *
 * The server now returns HTTP 409 Conflict when POST /api/functions is called
 * with a label that already exists. The client surfaces that as an exception
 * whose message contains both "status 409" and "already exists".
 */
private fun isAlreadyExistsError(e: Exception): Boolean {
    val msg = e.message ?: return false
    return msg.contains("status 409") || msg.contains("already exists")
}

private fun isNotFoundError(error: Throwable): Boolean =
    error.message?.let { it.contains("status 404") || it.contains("not found", ignoreCase = true) } == true

/**
 * Idempotent save: create the function, or PUT-update it if its label already
 * exists. Returns the function's encrypted ID either way (looked up by label on
 * the update path), so downstream get/update/delete-by-id keeps working.
 */
private suspend fun saveOrUpdate(
    client: EkoDBClient,
    func: UserFunction,
    functionIds: MutableSet<String>
): String {
    val id = try {
        client.saveFunction(func)
    } catch (e: Exception) {
        if (isAlreadyExistsError(e)) {
            // Label already exists — update the existing definition (PUT by label).
            client.updateFunction(func.label, func)
            println("ℹ️  Function '${func.label}' already existed — updated instead")
            // Resolve the encrypted ID by label so cleanup-by-id still works.
            client.getFunction(func.label).id
                ?: throw IllegalStateException("No ID returned for function '${func.label}'")
        } else {
            throw e
        }
    }
    functionIds += id
    return id
}

suspend fun setupTestData(client: EkoDBClient) {
    println("📋 Setting up test data...")

    try {
        client.deleteCollection(TEST_COLLECTION)
    } catch (error: Throwable) {
        if (!isNotFoundError(error)) throw error
    }

    for (i in 1..10) {
        val record = Record.new()
            .insert("name", "User $i")
            .insert("age", 20 + i)
            .insert("status", if (i % 2 == 0) "active" else "inactive")
            .insert("score", i * 10)
        client.insert(TEST_COLLECTION, record)
    }

    check(client.countDocuments(TEST_COLLECTION) == 10L) { "Expected exactly 10 test users" }

    println("✅ Test data ready\n")
}

suspend fun simpleQueryScript(client: EkoDBClient, functionIds: MutableSet<String>): String {
    println("📝 Example 1: Simple Query Function\n")

    val func = UserFunction(
        label = ACTIVE_LABEL,
        name = "Get Active Users",
        description = "Retrieve all active users",
        version = "1.0",
        parameters = emptyMap(),
        functions = listOf(
            FunctionStageConfig.Query(
                collection = TEST_COLLECTION,
                filter = queryCondition("status", QueryConditionOperator.Eq, JsonPrimitive("active"))
            )
        ),
        tags = listOf("users", "query")
    )

    val funcId = saveOrUpdate(client, func, functionIds)
    println("✅ Function saved: $funcId")

    val result = client.callFunction(ACTIVE_LABEL)
    check(result.records.size == 5) { "Expected 5 active users, got ${result.records.size}" }
    println("📊 Found ${result.records.size} records")
    println("⏱️  Execution time: ${result.stats.execution_time_ms}ms\n")
    
    return funcId
}

suspend fun parameterizedScript(client: EkoDBClient, functionIds: MutableSet<String>): String {
    println("📝 Example 2: Parameterized Function\n")

    val func = UserFunction(
        label = STATUS_LABEL,
        name = "Get Users By Status",
        version = "1.0",
        parameters = mapOf(
            "status" to ParameterDefinition(
                required = false,
                default = JsonPrimitive("active"),
                description = "Filter by user status"
            )
        ),
        functions = listOf(
            FunctionStageConfig.Query(
                collection = TEST_COLLECTION,
                filter = queryCondition("status", QueryConditionOperator.Eq, JsonPrimitive("{{status}}")),
                limit = 3
            )
        ),
        tags = listOf("users", "parameterized")
    )

    val funcId = saveOrUpdate(client, func, functionIds)
    println("✅ Function saved")

    val params = mapOf("status" to JsonPrimitive("active"))
    val result = client.callFunction(STATUS_LABEL, params)
    check(result.records.size == 3) { "Expected 3 limited users, got ${result.records.size}" }
    check(result.records.all { getValue<String>(it["status"]) == "active" }) {
        "Parameterized query returned a non-active user"
    }
    println("📊 Found ${result.records.size} users (limited)")
    println("⏱️  Execution time: ${result.stats.execution_time_ms}ms\n")
    return funcId
}

suspend fun aggregationScript(client: EkoDBClient, functionIds: MutableSet<String>): String {
    println("📝 Example 3: Aggregation Function\n")

    val func = UserFunction(
        label = STATS_LABEL,
        name = "User Statistics",
        version = "1.0",
        parameters = emptyMap(),
        functions = listOf(
            FunctionStageConfig.FindAll(collection = TEST_COLLECTION),
            FunctionStageConfig.Group(
                by_fields = listOf("status"),
                functions = listOf(
                    GroupFunctionConfig(
                        output_field = "count",
                        operation = GroupFunctionOp.Count
                    ),
                    GroupFunctionConfig(
                        output_field = "avg_score",
                        operation = GroupFunctionOp.Average,
                        input_field = "score"
                    )
                )
            )
        ),
        tags = listOf("analytics")
    )

    val funcId = saveOrUpdate(client, func, functionIds)
    println("✅ Function saved")

    val result = client.callFunction(STATS_LABEL)
    check(result.records.size == 2) { "Expected active and inactive groups" }
    val groupCounts = result.records.associate {
        requireNotNull(getValue<String>(it["status"])) to requireNotNull(getValue<Long>(it["count"]))
    }
    check(groupCounts == mapOf("active" to 5L, "inactive" to 5L)) {
        "Unexpected status counts: $groupCounts"
    }
    println("📊 Statistics: ${result.records.size} groups")
    result.records.forEach { record ->
        println("   $record")
    }
    println("⏱️  Execution time: ${result.stats.execution_time_ms}ms\n")
    
    return funcId
}

suspend fun functionManagement(client: EkoDBClient, getActiveUsersId: String, userStatsId: String) {
    println("📝 Example 4: function Management\n")

    val funcs = client.listFunctions()
    println("📋 Total functions: ${funcs.size}")

    // Get specific function (use encrypted ID)
    val func = client.getFunction(getActiveUsersId)
    println("🔍 Retrieved function: ${func.name}")

    // Update function (use encrypted ID)
    val updated = UserFunction(
        label = ACTIVE_LABEL,
        name = "Get Active Users (Updated)",
        description = "Updated description",
        version = "1.1",
        parameters = emptyMap(),
        functions = listOf(
            FunctionStageConfig.FindAll(collection = TEST_COLLECTION)
        ),
        tags = listOf("users")
    )
    client.updateFunction(getActiveUsersId, updated)
    println("✏️  function updated")

    // Delete function (use ID)
    client.deleteFunction(userStatsId)
    println("🗑️  function deleted")
    println()
    
    println("ℹ️  Note: GET/UPDATE/DELETE operations require the encrypted ID")
    println("ℹ️  Only CALL can use either ID or label\n")
}

suspend fun multiStageScript(client: EkoDBClient, functionIds: MutableSet<String>): String {
    println("📝 Example 5: Multi-Stage Pipeline\n")

    val func = UserFunction(
        label = TOP_LABEL,
        name = "Top Performing Users",
        version = "1.0",
        parameters = mapOf(
            "min_score" to ParameterDefinition(
                required = false,
                default = JsonPrimitive(50)
            )
        ),
        functions = listOf(
            FunctionStageConfig.Query(
                collection = TEST_COLLECTION,
                filter = queryCondition("score", QueryConditionOperator.Gte, JsonPrimitive("{{min_score}}"))
            ),
            FunctionStageConfig.Project(
                fields = listOf("name", "score", "status"),
                exclude = false
            )
        ),
        tags = listOf("analytics", "reporting")
    )

    val funcId = saveOrUpdate(client, func, functionIds)
    println("✅ Multi-stage function saved")

    val result = client.callFunction(TOP_LABEL, mapOf("min_score" to JsonPrimitive(50)))
    check(result.records.size == 6) { "Expected 6 users with score >= 50" }
    println("📊 Pipeline executed ${result.stats.stages_executed} stages")
    println("⏱️  Total execution time: ${result.stats.execution_time_ms}ms")
    println("📈 Stage breakdown:")
    result.stats.stage_stats.forEachIndexed { index, stage ->
        println("   ${index + 1}. ${stage.stage}: ${stage.execution_time_ms}ms (${stage.input_count} → ${stage.output_count} records)")
    }
    println()
    return funcId
}

suspend fun countScript(client: EkoDBClient, functionIds: MutableSet<String>): String {
    println("📝 Example 6: Count Users\n")

    val func = UserFunction(
        label = COUNT_LABEL,
        name = "Count All Users",
        version = "1.0",
        parameters = emptyMap(),
        functions = listOf(
            FunctionStageConfig.FindAll(collection = TEST_COLLECTION),
            FunctionStageConfig.Count(output_field = "count")
        ),
        tags = listOf("users", "count")
    )

    val funcId = saveOrUpdate(client, func, functionIds)
    println("✅ Count function saved")

    val result = client.callFunction(COUNT_LABEL)
    val count = result.records.firstOrNull()?.get("count")
    check(getValue<Long>(count) == 10L) { "Expected count function to return 10, got $count" }
    println("📊 Total user count: $count")
    println("⏱️  Execution time: ${result.stats.execution_time_ms}ms\n")
    return funcId
}

suspend fun cleanup(client: EkoDBClient, functionIds: Collection<String>): List<Throwable> {
    println("🧹 Cleaning up...")
    val errors = mutableListOf<Throwable>()
    var functionCleanupFailed = false
    for (id in functionIds) {
        try {
            client.deleteFunction(id)
        } catch (error: Throwable) {
            if (!isNotFoundError(error)) {
                functionCleanupFailed = true
                errors += error
            }
        }
    }
    if (functionIds.isNotEmpty() && !functionCleanupFailed) println("✅ Deleted test functions")
    try {
        client.deleteCollection(TEST_COLLECTION)
        println("✅ Deleted collection")
    } catch (error: Throwable) {
        if (!isNotFoundError(error)) errors += error
    }
    println()
    return errors
}

fun main() = runBlocking {
    println("🚀 ekoDB Functions Example (Kotlin Client)\n")

    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"
    val client = EkoDBClient.builder().baseUrl(baseUrl).apiKey(apiKey).build()
    val functionIds = linkedSetOf<String>()
    var failure: Throwable? = null

    try {
        println("✅ Client initialized\n")

        setupTestData(client)
        val getActiveUsersId = simpleQueryScript(client, functionIds)
        parameterizedScript(client, functionIds)
        val userStatsId = aggregationScript(client, functionIds)
        functionManagement(client, getActiveUsersId, userStatsId)
        functionIds.remove(userStatsId)
        multiStageScript(client, functionIds)
        countScript(client, functionIds)
    } catch (error: Throwable) {
        println("❌ Error: ${error.message}")
        failure = error
    } finally {
        for (cleanupError in cleanup(client, functionIds)) {
            failure = failure?.also { it.addSuppressed(cleanupError) } ?: cleanupError
        }
        try {
            client.close()
        } catch (cleanupError: Throwable) {
            failure = failure?.also { it.addSuppressed(cleanupError) } ?: cleanupError
        }
    }
    failure?.let { throw it }
    println("✅ All examples completed successfully!")
    println("\n💡 Key Advantages of Using the Client:")
    println("   • Automatic token management")
    println("   • Type-safe Stage builders")
    println("   • Built-in error handling")
}

File: examples/rust/examples/client_functions.rs
///! Functions Example for ekoDB Rust Client
///!
///! Demonstrates creating, managing, and executing functions
use ekodb_client::{
    Client, FieldType, Function, GroupFunctionConfig, GroupFunctionOp, ParameterDefinition,
    QueryConditionOperator, QueryExpression, Record, SortFieldConfig, UserFunction,
};
use std::{collections::HashMap, env};

const TEST_COLLECTION: &str = "functions_users_client_rs";
const ACTIVE_LABEL: &str = "get_active_users_client_rs";
const UPDATED_ACTIVE_LABEL: &str = "get_active_users_updated_client_rs";
const STATUS_LABEL: &str = "get_users_by_status_client_rs";
const STATS_LABEL: &str = "user_stats_client_rs";

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Create ekoDB client
    let client = Client::builder()
        .base_url(&base_url)
        .api_key(&api_key)
        .build()?;

    println!("=== ekoDB Rust Client - Functions Example ===\n");

    // Setup test data
    let _ = client.delete_collection(TEST_COLLECTION).await;
    let operation_result: Result<(), Box<dyn std::error::Error>> = async {
        setup_test_data(&client).await?;
        let get_active_users_id = simple_query_script(&client).await?;
        let get_users_by_status_id = parameterized_script(&client).await?;
        let user_stats_id = aggregation_script(&client).await?;
        script_management(
            &client,
            &get_active_users_id,
            &get_users_by_status_id,
            &user_stats_id,
        )
        .await?;
        Ok(())
    }
    .await;

    let mut cleanup_errors = Vec::new();
    for label in [
        ACTIVE_LABEL,
        UPDATED_ACTIVE_LABEL,
        STATUS_LABEL,
        STATS_LABEL,
    ] {
        if let Err(error) = client.delete_function(label).await {
            if !error.to_string().contains("not found") {
                cleanup_errors.push(format!("function {label}: {error}"));
            }
        }
    }
    if let Err(error) = client.delete_collection(TEST_COLLECTION).await {
        cleanup_errors.push(format!("collection {TEST_COLLECTION}: {error}"));
    }

    if let Err(error) = operation_result {
        if !cleanup_errors.is_empty() {
            eprintln!("Cleanup also failed: {}", cleanup_errors.join("; "));
        }
        return Err(error);
    }
    if !cleanup_errors.is_empty() {
        return Err(format!("Cleanup failed: {}", cleanup_errors.join("; ")).into());
    }
    println!("\n✅ All examples completed!");
    Ok(())
}

/// Save a function idempotently: if the label already exists (HTTP 409),
/// update the existing definition instead, then return its id so the rest of
/// the example (get/update/delete by id) continues to work unchanged.
async fn save_or_update(
    client: &Client,
    function: UserFunction,
) -> Result<String, Box<dyn std::error::Error>> {
    let label = function.label.clone();
    match client.save_function(function.clone()).await {
        Ok(id) => Ok(id),
        Err(ekodb_client::Error::Api { code: 409, .. }) => {
            client.update_function(&label, function).await?;
            println!("ℹ️  Function '{}' already existed — updated instead", label);
            // We don't get a fresh id back from update; fetch by label to recover it.
            let existing = client.get_function(&label).await?;
            Ok(existing.id.unwrap_or(label))
        }
        Err(e) => Err(Box::new(e)),
    }
}

async fn setup_test_data(client: &Client) -> Result<(), Box<dyn std::error::Error>> {
    println!("📋 Setting up test data...");

    for i in 1..=10 {
        let mut record = Record::new();
        record.insert("name", FieldType::String(format!("User {}", i)));
        record.insert("age", FieldType::Integer(20 + i));
        record.insert(
            "status",
            FieldType::String(if i % 2 == 0 { "active" } else { "inactive" }.to_string()),
        );
        record.insert("score", FieldType::Integer(i * 10));
        client.insert(TEST_COLLECTION, record, None).await?;
    }

    println!("✅ Test data ready\n");
    Ok(())
}

async fn simple_query_script(client: &Client) -> Result<String, Box<dyn std::error::Error>> {
    println!("📝 Example 1: Simple Query Function\n");

    let script =
        UserFunction::new(ACTIVE_LABEL, "Get Active Users").with_function(Function::Query {
            collection: TEST_COLLECTION.to_string(),
            filter: Some(QueryExpression::condition(
                "status",
                QueryConditionOperator::Eq,
                serde_json::json!("active"),
            )),
            sort: None,
            limit: None,
            skip: None,
        });

    let id = save_or_update(client, script).await?;
    println!("✅ Function saved: {}", id);

    let result = client.call_function(ACTIVE_LABEL, None).await?;
    println!("📊 Found {} active users\n", result.records.len());

    Ok(id)
}

async fn parameterized_script(client: &Client) -> Result<String, Box<dyn std::error::Error>> {
    println!("📝 Example 2: Parameterized Function\n");

    let script = UserFunction::new(STATUS_LABEL, "Get Users By Status")
        .with_parameter(
            ParameterDefinition::new("status")
                .with_default(FieldType::String("active".to_string())),
        )
        .with_parameter(ParameterDefinition::new("limit").with_default(FieldType::Integer(10)))
        .with_function(Function::Query {
            collection: TEST_COLLECTION.to_string(),
            filter: Some(QueryExpression::condition(
                "status",
                QueryConditionOperator::Eq,
                serde_json::json!("{{status}}"),
            )),
            sort: Some(vec![SortFieldConfig::new("score").descending()]),
            limit: Some(serde_json::json!("{{limit}}")),
            skip: None,
        });

    let id = save_or_update(client, script).await?;
    println!("✅ Function saved: {}", id);

    let mut params = HashMap::new();
    params.insert(
        "status".to_string(),
        FieldType::String("active".to_string()),
    );
    params.insert("limit".to_string(), FieldType::Integer(3));

    let result = client.call_function(STATUS_LABEL, Some(params)).await?;
    println!("📊 Found {} users (limited)\n", result.records.len());

    Ok(id)
}

async fn aggregation_script(client: &Client) -> Result<String, Box<dyn std::error::Error>> {
    println!("📝 Example 3: Aggregation Function\n");

    let script = UserFunction::new(STATS_LABEL, "User Statistics")
        .with_function(Function::FindAll {
            collection: TEST_COLLECTION.to_string(),
        })
        .with_function(Function::Group {
            by_fields: vec!["status".to_string()],
            functions: vec![
                GroupFunctionConfig::new("count", GroupFunctionOp::Count),
                GroupFunctionConfig::new("avg_score", GroupFunctionOp::Average)
                    .with_input_field("score"),
            ],
        });

    let id = save_or_update(client, script).await?;
    println!("✅ Function saved: {}", id);

    let result = client.call_function(STATS_LABEL, None).await?;
    println!("📊 Statistics: {} groups\n", result.records.len());

    Ok(id)
}

async fn script_management(
    client: &Client,
    get_active_users_id: &str,
    _get_users_by_status_id: &str,
    user_stats_id: &str,
) -> Result<(), Box<dyn std::error::Error>> {
    println!("📝 Example 4: Function Management\n");

    // List all functions
    let functions = client.list_functions(None).await?;
    println!("📋 Total functions: {}", functions.len());

    // Get specific function by ID
    let function = client.get_function(get_active_users_id).await?;
    println!("🔍 Retrieved function: {}", function.name);

    // Update function by ID
    let updated = UserFunction::new(UPDATED_ACTIVE_LABEL, "Get Active Users (Updated)")
        .with_description("Updated description")
        .with_version("1.1")
        .with_function(Function::FindAll {
            collection: TEST_COLLECTION.to_string(),
        })
        .with_tag("users");

    client.update_function(get_active_users_id, updated).await?;
    println!("✏️  Function updated");

    // Delete function by ID
    client.delete_function(user_stats_id).await?;
    println!("🗑️  Function deleted\n");

    println!("ℹ️  Note: GET/UPDATE/DELETE use IDs. Only CALL supports labels.");

    println!();

    Ok(())
}
